homeassistant/components/media_player/mediaroom.py

```diff
@@ -9,134 +9,169 @@
         """Initialize the device."""
         from pymediaroom import Remote
 
-        self.stb = Remote(host, timeout=timeout)
-        _LOGGER.info(
-            "Found %s at %s%s", name, host,
-            " - I'm optimistic" if optimistic else "")
-        self._name = name
-        self._is_standby = not optimistic
-        self._current = None
+        self.host = host
+        self.stb = Remote(host)
+        _LOGGER.info("Found STB at %s%s", host,
+                     " - I'm optimistic" if optimistic else "")
+        self._channel = None
         self._optimistic = optimistic
-        self._state = STATE_STANDBY
+        self._state = STATE_PLAYING if optimistic else STATE_STANDBY
+        self._name = '{}_{}'.format(DOMAIN, device_id)
+        if device_id:
+            self._unique_id = device_id
+        else:
+            self._unique_id = self.stb.device_id
 
-    def update(self):
+    @property
+    def should_poll(self):
+        """No polling needed."""
+        return False
+
+    async def async_added_to_hass(self):
         """Retrieve latest state."""
-        if not self._optimistic:
-            self._is_standby = self.stb.get_standby()
-        if self._is_standby:
-            self._state = STATE_STANDBY
-        elif self._state not in [STATE_PLAYING, STATE_PAUSED]:
-            self._state = STATE_PLAYING
-        _LOGGER.debug(
-            "%s(%s) is [%s]",
-            self._name, self.stb.stb_ip, self._state)
-
-    def play_media(self, media_type, media_id, **kwargs):
+        def notify_received(notify):
+            """Process STB state from NOTIFY message."""
+            self._state = MediaroomDevice.STATE_MAP[
+                self.stb.notify_callback(notify)]
+            _LOGGER.debug("STB(%s) is [%s]", self.host, self._state)
+            self.schedule_update_ha_state()
+
+        async_dispatcher_connect(self.hass, SIGNAL_STB_NOTIFY,
+                                 notify_received)
+
+    async def async_play_media(self, media_type, media_id, **kwargs):
         """Play media."""
-        _LOGGER.debug(
-            "%s(%s) Play media: %s (%s)",
-            self._name, self.stb.stb_ip, media_id, media_type)
-        if media_type != MEDIA_TYPE_CHANNEL:
-            _LOGGER.error('invalid media type')
-            return
-        if media_id.isdigit():
-            media_id = int(media_id)
-        else:
-            return
-        self.stb.send_cmd(media_id)
-        self._state = STATE_PLAYING
+        from pymediaroom import PyMediaroomError
+        try:
+            _LOGGER.debug("STB(%s) Play media: %s (%s)", self.stb.stb_ip,
+                          media_id, media_type)
+            if media_type != MEDIA_TYPE_CHANNEL:
+                _LOGGER.error('invalid media type')
+                return
+            if not media_id.isdigit():
+                _LOGGER.error("media_id must be a channel number")
+                return
+
+            await self.stb.send_cmd(int(media_id))
```

Comment:
It seems like over in pymediaroom you're creating a new connection *for every single command* you send. Isn't that incredibly inefficient or is there no other way?

https://github.com/dgomes/pymediaroom/blob/2840407ce254c17ad934f9c4e21188ea6b1a398f/pymediaroom/remote.py#L75-L77

Reply:
it's actually very important :)

The mediaroom STB only accepts 1 open socket, if I keep the socket open it will not accept connections from the official app.